core/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java

```diff
@@ -131,4 +145,180 @@ public void writeTo(StreamOutput out) throws IOException {
+                            if (!allowExplicitIndex) {
+                                throw new IllegalArgumentException("explicit index in multi search is not allowed");
+                            }
+                            searchRequest.indices(nodeStringArrayValue(value));
+                        } else if ("type".equals(entry.getKey()) || "types".equals(entry.getKey())) {
+                            searchRequest.types(nodeStringArrayValue(value));
+                        } else if ("search_type".equals(entry.getKey()) || "searchType".equals(entry.getKey())) {
+                            searchRequest.searchType(nodeStringValue(value, null));
+                        } else if ("request_cache".equals(entry.getKey()) || "requestCache".equals(entry.getKey())) {
+                            searchRequest.requestCache(nodeBooleanValue(value, entry.getKey()));
+                        } else if ("preference".equals(entry.getKey())) {
+                            searchRequest.preference(nodeStringValue(value, null));
+                        } else if ("routing".equals(entry.getKey())) {
+                            searchRequest.routing(nodeStringValue(value, null));
+                        }
+                    }
+                    defaultOptions = IndicesOptions.fromMap(source, defaultOptions);
+                }
+            }
+            searchRequest.indicesOptions(defaultOptions);
+
+            // move pointers
+            from = nextMarker + 1;
+            // now for the body
+            nextMarker = findNextMarker(marker, from, data, length);
+            if (nextMarker == -1) {
+                break;
+            }
+            BytesReference bytes = data.slice(from, nextMarker - from);
+            try (XContentParser parser = xContent.createParser(registry, bytes)) {
+                consumer.accept(searchRequest, parser);
+            }
+            // move pointers
+            from = nextMarker + 1;
+        }
+    }
+
+    private static int findNextMarker(byte marker, int from, BytesReference data, int length) {
+        for (int i = from; i < length; i++) {
+            if (data.get(i) == marker) {
+                return i;
+            }
+        }
+        if (from != length) {
+            throw new IllegalArgumentException("The msearch request must be terminated by a newline [\n]");
+        }
+        return -1;
+    }
+
+    public static BytesRef writeMultiLineFormat(MultiSearchRequest multiSearchRequest, XContent xContent) throws IOException {
+        BytesRefBuilder builder = new BytesRefBuilder();
+        for (SearchRequest request : multiSearchRequest.requests()) {
+            try (XContentBuilder xContentBuilder = XContentBuilder.builder(xContent)) {
+                xContentBuilder.startObject();
+                if (request.indices() != null) {
+                    xContentBuilder.field("indices", request.indices());
```

Comment:
nit: could we use `index` rather than `indices`? That is what we have in our docs, indices is just a synonym but I don't think that other clients use it either.